staging_vespalib/src/vespa/vespalib/util/clock.cpp

```diff
@@ -1,24 +1,79 @@
 // Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
 
 #include "clock.h"
+#include <vespa/fastos/thread.h>
+#include <mutex>
+#include <condition_variable>
 #include <cassert>
 #include <chrono>
 
 namespace vespalib {
 
+namespace clock::internal {
+
+class Updater : public FastOS_Runnable
+{
+private:
+    Clock                   & _clock;
+    int                       _timePeriodMS;
+    std::mutex                _lock;
+    std::condition_variable   _cond;
+    bool                      _stop;
+
+
+    void Run(FastOS_ThreadInterface *thisThread, void *arguments) override;
+
+public:
+    Updater(Clock & clock, double timePeriod=0.100);
+    ~Updater();
+
+    void stop();
+};
+
+Updater::Updater(Clock & clock, double timePeriod)
+    : _clock(clock),
+      _timePeriodMS(static_cast<uint32_t>(timePeriod*1000)),
+      _lock(),
+      _cond(),
+      _stop(false)
+{ }
+
+Updater::~Updater() = default;
+
+void
+Updater::Run(FastOS_ThreadInterface *thread, void *arguments)
+{
+    (void) arguments;
+    _clock._running = true;
+    std::unique_lock<std::mutex> guard(_lock);
+    while ( ! thread->GetBreakFlag() && !_stop) {
+        _clock.setTime();
+        _cond.wait_for(guard, std::chrono::milliseconds(_timePeriodMS));
+    }
+    _clock._running = false;
+}
+
+void
+Updater::stop()
+{
+    std::lock_guard<std::mutex> guard(_lock);
+    _stop = true;
+    _cond.notify_all();
+}
+
+}
+
 Clock::Clock(double timePeriod) :
      _timeNS(0u),
-     _timePeriodMS(static_cast<uint32_t>(timePeriod*1000)),
-     _lock(),
-     _cond(),
-     _stop(false),
+     _updater(std::make_unique<clock::internal::Updater>(*this, timePeriod)),
      _running(false)
 {
     setTime();
 }
 
 Clock::~Clock()
 {
+    _updater.reset();
```

Comment:
Is it possible for this to race with the clock updater thread?
1. `Updater::stop()` is called while clock thread is waiting on the condition variable and is not holding the mutex
2. `Updater::stop()` returns before the clock thread has woken up to exit its loop
3. `_updater` is deleted, causing `Updater::Run()` to poke at freed memory

Reply:
That is correct. I have added an explicit join if necessary.
The only use had protection on the outside due a threadPool.Close.